packages/lesswrong/server/fmCrosspost/crosspost.ts

```
@@ -1,31 +1,43 @@
+import pick from 'lodash/pick'
 import Users from "../../lib/collections/users/collection";
 import { randomId } from "../../lib/random";
+import { loggerConstructor } from "../../lib/utils/logging";
+import { UpdateCallbackProperties } from "../mutationCallbacks";
 import { DenormalizedCrosspostData, denormalizedFieldKeys, extractDenormalizedData } from "./denormalizedFields";
 import { makeCrossSiteRequest } from "./resolvers";
 import { signToken } from "./tokens";
-import { Crosspost, CrosspostPayload, UpdateCrosspostPayload } from "./types";
+import type { Crosspost, CrosspostPayload, UpdateCrosspostPayload } from "./types";
 
-export const performCrosspost = async <T extends Crosspost>(post: T): Promise<T> => {
+export async function performCrosspost<T extends Crosspost>(post: T): Promise<T> {
+  const logger = loggerConstructor('callbacks-posts')
+  logger('performCrosspost()')
+  logger('post info:', pick(post, ['title', 'fmCrosspost']))
+  // TODO; validate userId owns foreignPost && currentUser === userId || currentUser.isAdmin
   if (!post.fmCrosspost || !post.userId || post.draft) {
+    logger('post is not a crosspost or is a draft, returning')
     return post;
   }
 
   const {isCrosspost, hostedHere, foreignPostId} = post.fmCrosspost;
   if (!isCrosspost || !hostedHere || foreignPostId) {
+    logger ('post is not a crosspost, or is a foreign-hosted crosspost, or has already been crossposted, returning')
     return post;
   }
 
   if (post.isEvent) {
+    logger ('post is an event, throwing')
     throw new Error("Events cannot be crossposted");
   }
 
   const user = await Users.findOne({_id: post.userId});
   if (!user || !user.fmCrosspostUserId) {
+    logger('user has not connected an account, throwing')
```

I don't really see the point of logging just before throwing an error

---

Because sometimes errors are swallowed. ¯\\\_(ツ)_/¯